README.md
# AWS LS
Command line tool that returns information over all S3 buckets in a Amazon account

* must work on linux, OSX and Windows
* easy to install and use
* less frameworks, no external tools allowed
* must be fast!

# Installation

1. Install [Python](https://www.python.org/downloads/) 3.6 (or later) on your system
2. On a terminal at downloaded folder:
```
pip install -r requirements.txt
```
3. Verify your Amazon Credentials in **~/.aws/credentials** file
    * Make sure to have [Cost Explorer](https://console.aws.amazon.com/cost-reports/home) enabled (it take at least 24h after activation to be available)
    * Make sure to have enabled the following API access in [IAM](https://console.aws.amazon.com/iam/home):
      - AmazonS3ReadOnlyAccess
      - AdministratorAccess (for cost)   
```
[default]
aws_access_key_id=AKIAIOSFODNN7EXAMPLE
aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

# Usage
```
python awsls.py 
```
## Arguments
```
  --help, -h
  --verbose, -v
  --bucketfilter, -b       BUCKETNAME
  --filefilter, -f         FILENAME
  --regionfilter, -r       {us-east-2,us-east-1,us-west-1,us-west-2,ap-east-1,ap-south-1,...}
  --unit, -u               {kb,KB,mb,MB,gb,GB,TB}
  --sort, -s               {region,creation,files,modified,size,cost}
  --type, -t               {STANDARD,REDUCED_REDUNDANCY,STANDARD_IA,ONEZONE_IA,INTELLIGENT_TIE}
  --group, -g
```
## Examples
#### Get information from all buckets, sort by cost
```
python awsls.py -s cost
```
<details>
  <summary>Result</summary>  
  
```
                   Region       Creation Date  Files             Size  % Size       Last modified      Cost
mybucketprd002  us-east-2 2019-09-13 12:05:44    373  721070816 bytes  86.03% 2019-09-14 15:01:38  1.193732
mybucketprd003  sa-east-1 2019-09-13 12:06:21      3  113500480 bytes  13.54% 2019-09-14 19:49:03  0.187900
mybucketprd001  us-east-2 2019-09-13 12:05:21      8    3552303 bytes   0.42% 2019-09-15 21:07:33  0.005881
bucket1         us-east-2 2019-09-15 19:26:16      1          0 bytes   0.00% 2019-09-15 19:27:18  0.000000
```
</details>

#### Get details from a specific bucket named "mybucket1"
```
python awsls.py -b bucket1
```
<details>
  <summary>Result</summary>  
  
```
            Region       Creation Date  Files     Size % Size       Last modified  Cost
bucket1  us-east-2 2019-09-15 19:26:16      1  0 bytes  0.00% 2019-09-15 19:27:18     0
```
</details>

#### Get information from buckets named "prd", sort by modified date, display results in MB
```
python awsls.py -b prd -s modified -u MB
```
<details>
  <summary>Result</summary>  
  
```
                   Region       Creation Date  Files    Size  % Size       Last modified      Cost
mybucketprd001  us-east-2 2019-09-13 12:05:21      8    3 MB   0.42% 2019-09-15 21:07:33  0.005881
mybucketprd003  sa-east-1 2019-09-13 12:06:21      3  108 MB  13.54% 2019-09-14 19:49:03  0.187900
mybucketprd002  us-east-2 2019-09-13 12:05:44    373  688 MB  86.03% 2019-09-14 15:01:38  1.193732
```
</details>

#### Get information from buckets in region us-east-2
```
python awsls.py -r us-east-2
```
<details>
  <summary>Result</summary>  
  
```
                   Region       Creation Date  Files             Size  % Size       Last modified      Cost
bucket1         us-east-2 2019-09-15 19:26:16      1          0 bytes   0.00% 2019-09-15 19:27:18  0.000000
mybucketprd001  us-east-2 2019-09-13 12:05:21      8    3552303 bytes   0.49% 2019-09-15 21:07:33  0.006802
mybucketprd002  us-east-2 2019-09-13 12:05:44    373  721070816 bytes  99.51% 2019-09-14 15:01:38  1.380711
```
</details>

#### Get information from all buckets, group by region, sum cost and size
```
python awsls.py -g
```
<details>
  <summary>Result</summary>  
  
```
           Files      Cost
Region
sa-east-1      3  0.187900
us-east-2    382  1.199613
```
</details>

#### Get information from all buckets, filter files that contains CACHE in their path
```
python awsls.py -f CACHE
```
<details>
  <summary>Result</summary>  
  
```
                   Region       Creation Date  Files           Size   % Size        Last modified      Cost
bucket1         us-east-2 2019-09-15 19:26:16      1        0 bytes    0.00%  2019-09-15 19:27:18  0.000000
mybucketprd001  us-east-2 2019-09-13 12:05:21      6  1974110 bytes  100.00%  2019-09-15 21:07:33  1.387513
mybucketprd002  us-east-2 2019-09-13 12:05:44      0        0 bytes    0.00%                    -  0.000000
mybucketprd003  sa-east-1 2019-09-13 12:06:21      0        0 bytes    0.00%                    -  0.000000
```
</details>

#### Get information from buckets named "mybucketprd2", filter by STANDARD storage class, display results in KB
```
python awsls.py -b mybucketprd2 -t STANDARD -u KB
```
<details>
  <summary>Result</summary>  
  
```
                   Region       Creation Date  Files        Size   % Size       Last modified      Cost
mybucketprd002  us-east-2 2019-09-13 12:05:44    373  704,171 KB  100.00% 2019-09-14 15:01:38  1.387513
```
</details>


changelog.txt
Day 4
* Documentation and code refinement
* Bug fixes

Day 3
* Refactored code, performance improvements
* Sort capabilities
* More filters options
* Arguments improvements
* Cost finally available (thank you Amazon)
* Basic code documentation
 
Day 2
* Added basic data, such as get bucket name, region, objects, etc
* Command-line arguments done properly
* Size formating using bit shift operator
* Added bucket name, storage type and others filters
* Group by region option

Day 1
* Project requirements review
* Study of AWS SDK and it's intregations
* Explored possibilities (Python, Go, SDK vs web requests)

Day 0
* Received project
* Create Amazon AWS account


awslsmp.py
# Amazon SDK
import boto3

# Amazon SDK exceptions
from boto3.exceptions import botocore

# Arguments parsing utility
import argparse

# System library to exit to console
import sys

# Datetime for date manipulation
from datetime import datetime, timedelta

# Pandas for data filtering and visualization
import pandas as pd

# Parallelizing
import multiprocessing as mp

class cBucket:
    """ Class that represents a bucket in AWS. Contains name, creation date, files count, total files size, most recent file date modification, files count per storage types, cost and region
    """
    # total_size_list = []

    def __init__(self, name):
        self.name = name
        self.creation_date = datetime(1, 1, 1).replace(tzinfo=None)
        self.files_count = 0
        self.files_size = 0
        self.last_modified = datetime(1, 1, 1).replace(tzinfo=None)
        self.storage_types = {}
        self.cost = 0.
        self.region = ''

def format_size(value):
    """ Convert bytes to other units, such as KB, MB, GB
    """
    bit_shift = {"B": 0,
            "kb": 7,
            "KB": 10,
            "mb": 17,
            "MB": 20,
            "gb": 27,
            "GB": 30,
            "TB": 40,}
    if args.unit in bit_shift:
        return "{:,.0f} {}".format(value / float(1 << bit_shift[args.unit]),args.unit) 
    return "{} bytes".format(value)

def print_buckets(buckets):
    """ Print bucket in different ways
    """
    data = []
    names = []    
    for bucket in buckets:
        # total_size = sum(bucket.total_size_list)
        total_size = 4000000#sum(bucket.total_size_list)
        if total_size == 0:
            bucket.cost = 0
        else:
            bucket.cost = (bucket.files_size/total_size) * (float(total_cost))

        d = [            
            bucket.region,
            bucket.creation_date,
            bucket.files_count,
            format_size(bucket.files_size),
            bucket.last_modified,
            bucket.cost ]
        data.append(d)
        names.append(bucket.name)

    features = {'region':'Region', 'creation':'Creation Date', 'files':'Files', 'size':'Size', 'modified':'Last modified' , 'cost':'Cost'}
    dataframe = pd.DataFrame(data=data,columns=features.values(),index=names)

    if args.group:
        dataframe = dataframe.groupby('Region').sum()        
    elif args.region:
        dataframe = dataframe[dataframe.Region == args.region]
    
    if args.sort:
        dataframe = dataframe.sort_values(features[args.sort], ascending=False)

    if not dataframe.empty:
        print(dataframe)
    else:
        print("No buckets match your criteria")

def isValidFile(f):
    """ Return a boolean indicating if file f is valid
    """
    result = False
    if f.storage_class == args.type or not(args.type):
        if str(args.filter) in str(f.key) or not(args.filter):
            result = True
    return result
    
def get_bucket_details(bucket):
    """ Get bucket details from amazon, create an populate a cBucket object and return it at the end
    """ 
    if args.verbose:        
            print('\tRetreving files in bucket {}'.format(bucket.name), end='...', flush=True)
    files = bucket.objects.all()
    if args.verbose:        
            print('done')
    
    last_modified_file = datetime(1, 1, 1).replace(tzinfo=None)
    total_files_size = 0
    valid_files_count = 0
    storage_types = {}

    for f in files:
        if isValidFile(f):
            total_files_size+=f.size
            valid_files_count+=1

            last_change = f.last_modified.replace(tzinfo=None)
            if last_modified_file < last_change or last_modified_file=='':
                last_modified_file = last_change
            
            if (f.storage_class in storage_types): 
                storage_types[f.storage_class] += 1
            else: 
                storage_types[f.storage_class] = 1

    location_response = client.get_bucket_location( Bucket=bucket.name )
    location = location_response['LocationConstraint']

    coveo_bucket = cBucket(bucket.name)
    coveo_bucket.creation_date = bucket.creation_date.replace(tzinfo=None)
    coveo_bucket.files_count = valid_files_count
    coveo_bucket.files_size = total_files_size
    coveo_bucket.last_modified = last_modified_file.replace(tzinfo=None)
    coveo_bucket.storage_types = storage_types   
    coveo_bucket.region = location_response['LocationConstraint']
    # coveo_bucket.total_size_list.append(total_files_size)

    return coveo_bucket

def get_y_n(text):
    """ Get Y or N from user with a friendly message. No input returns False (N)
    """
    valid_answers = [ 'y','n','Y','N', '' ]
    while True:
        data = input(text)
        if data not in valid_answers:
            continue
        elif data=='y' or data=='Y':
            return True
        else:
            return False

def get_buckets(s3):
    """ Get all buckets from Amazon and return a filtered subset of them using command-line filters
    """
    buckets = s3.buckets.all()
    filtered_buckets = []
    
    for bucket in buckets:
        if str(args.bucketfilter) in str(bucket.name) or not(args.bucketfilter):
            location = ''
            if args.region:
                location_response = client.get_bucket_location( Bucket=bucket.name )
                location = location_response['LocationConstraint']
            if location == args.region or not(args.region):
                filtered_buckets.append(bucket)
    
    return filtered_buckets, len(filtered_buckets)

def parsearguments():
    """ Parse arguments from command-line
    """
    parser = argparse.ArgumentParser()
    help_text = 'This tool returns useful information from AWS S3 bucket'
    parser = argparse.ArgumentParser(description = help_text)
    
    aws_regions = [ 'us-east-2',
        'us-east-1',
        'us-west-1',
        'us-west-2',
        'ap-east-1',
        'ap-south-1',
        'ap-northeast-3',
        'ap-northeast-2',
        'ap-southeast-1',
        'ap-southeast-2',
        'ap-northeast-1',
        'ca-central-1',
        'cn-north-1',
        'cn-northwest-1',
        'eu-central-1',
        'eu-west-1',
        'eu-west-2',
        'eu-west-3',
        'eu-north-1',
        'me-south-1',
        'sa-east-1',
        'us-gov-east-1',
        'us-gov-west-1']

    parser.add_argument("--verbose", "-v", help="Verbose mode", action="store_true")
    parser.add_argument("--group", "-g", help="Group by regions", action="store_true")    
    # parser.add_argument("--cost", "-c", help="Try to get the cost", action="store_true")    
    parser.add_argument("--bucketfilter", "-b", help="Bucket name filter" )
    parser.add_argument("--region", "-r", help="Region filter", choices=aws_regions)
    parser.add_argument("--filter", "-f", help="File filter")
    parser.add_argument("--type", "-t", help="Storage type", choices=[ 'STANDARD', 'REDUCED_REDUNDANCY', 'STANDARD_IA', 'ONEZONE_IA', 'INTELLIGENT_TIE' ] )
    parser.add_argument("--unit", "-u", help="Size unit", choices=[ 'kb', 'KB', 'mb', 'MB', 'gb', 'GB', 'TB'])
    parser.add_argument("--sort", "-s", help="Sort by feature (descending)", choices=[ 'region', 'creation', 'files', 'modified', 'size', 'cost' ])

    return parser.parse_args()

def getTotalCost():
    """ Get total storage cost from last day from Amazon. Can't be filtered
    """
    if args.bucketfilter or args.type or args.filter:
        print("WARNING: cost estimation isn't compatible with filtering. Cost results may not be reliabe.")

    today = datetime.today().date()
    yesterday = today-timedelta(days=2)
    ce = boto3.client('ce')
    cost_response = ce.get_cost_and_usage(TimePeriod={
                                            'Start': str(yesterday),
                                            'End': str(today)
                                        },
                                        Granularity='MONTHLY',
                                        Metrics=[ 'AmortizedCost']
                                    )
    c = cost_response['ResultsByTime'].pop()['Total']['AmortizedCost']['Amount']
    return c

def main():
    """ Main method that control global flow
    """
    try:
        buckets, buckets_count = get_buckets(s3)
        
        global total_cost
        total_cost = getTotalCost()

        if args.verbose:        
            print('{} buckets found'.format(buckets_count))
            
        if buckets_count > HIGH_BUCKETS_COUNT_WARNING:
            allowed = get_y_n("WARNING: high number of buckets found ({}), it may take a while to get all data. Continue (y/N)?".format(buckets_count))
            if not(allowed):
                sys.exit()
    
        pool = mp.Pool(mp.cpu_count())
        target_buckets = []
        target_buckets = [ pool.apply(get_bucket_details, args=(bucket)) for bucket in buckets ]
        pool.close()
        # for bucket in buckets:  
                
        print_buckets(target_buckets)

    except botocore.exceptions.NoCredentialsError:
        print("Sorry, I can't find your credentials file. Please double check the ~/.aws/credentials configuration file")
    except botocore.exceptions.ClientError as e:
        print("Client error, see details below:\n {}".format(e))
    except Exception as e:
        print("Unexpected error occoured, more details bellow\n{}".format(e))

# Set constants
HIGH_BUCKETS_COUNT_WARNING = 3

# Initialize global variables
total_cost = 0

# Parse command-line arguments
args = parsearguments()

# Instantiates Amazon SDK resources
s3 = boto3.resource('s3')
client = boto3.client('s3')

main()

aws3inspector.py
# Amazon SDK
import boto3

# Amazon SDK exceptions
from boto3.exceptions import botocore

# Arguments parsing utility
import argparse

# System library to exit to console
import sys

# others, for debug
import time, random

from datetime import datetime

def pretty_bytes(value):
    if args.unit == "kb":
        return '{:.2f} KB'.format(value/float(1<<10))

    if args.unit == "mb":
        return '{:.2f} MB'.format(value/float(1<<20))

    return "{} bytes".format(value)

def print_bucket(bucket):    
    if args.verbose:        
            print('\tRetreving files in bucket {}'.format(bucket.name), end='...', flush=True)

    files = bucket.objects.all()
    # time.sleep(random.randrange(1,4))

    if args.verbose:        
            print('done')

    total_files_size = 0
    files_count = len(list(files))
    last_modified_file = datetime(1900, 9, 1).replace(tzinfo=None)
    storage_types = {}

    for file in files: 
        # print(file.key, '{:,.0f}'.format(file.size/float(1<<20))+' MB') #, file.last_modified)
        total_files_size+=file.size

        last_change = file.last_modified.replace(tzinfo=None)
        if last_modified_file < last_change:
            last_modified_file = last_change
        
        if (file.storage_class in storage_types): 
            storage_types[file.storage_class] += 1
        else: 
            storage_types[file.storage_class] = 1
        # storage_types[file.storage_class] = storage_types[file.storage_class] + 1
        
        # obj.last_modified

    print("Bucket name:{}".format(bucket.name))
    # print("Bucket location:{}".format(bucket.))
    print("Total file size: {}".format(pretty_bytes(total_files_size)))
    print("Number of files: {}".format(files_count))
    print("Last modified file: {}".format(last_modified_file))
    print("Bucket creation date: {}".format(bucket.creation_date))
    print("Storage type: {}".format(storage_types))
    

def print_buckets_info(buckets):
    for bucket in buckets:
        print("Bucket name:{}\nBucket creation date:{}".format(bucket.name, bucket.creation_date))

def get_buckets(s3):
    if args.bucket:
        return [ s3.Bucket(args.bucket) ]
    return s3.buckets.all()

def get_y_n(text):
    valid_answers = [ 'y','n','Y','N', '' ]
    while True:
        data = input(text)
        if data not in valid_answers:
            continue
        elif data=='y' or data=='Y':
            return True
        else:
            return False

def main():
    try:


        buckets =  get_buckets(s3)
        buckets_count = len(list(buckets))

        # print_bucket
        # s_info(buckets)

        if args.verbose:        
            print('{} buckets found'.format(buckets_count))

        if buckets_count > HIGH_BUCKETS_COUNT_WARNING:
            allowed = get_y_n("WARNING: high number of buckets found, may take sometime to get all data. Continue (y/N)?")
            if not(allowed):
                sys.exit()
    
        for bucket in buckets:  
            response = client.get_bucket_location( Bucket=bucket.name )
            bucket_location = response['LocationConstraint']
            print("Bucket location: {}".format(bucket_location))

            print_bucket(bucket)
         

    except botocore.exceptions.NoCredentialsError:
        print("Sorry, I can't find your credentials file. Please double check the ~/.aws/credentials configuration file")
    except botocore.exceptions.ClientError as e:
        print("Client error, see details below:\n {}".format(e))
    except Exception as e:
        print("Unexpected error occoured, more details bellow\n{}".format(e))


def parsearguments():        
    # initiate the parser
    parser = argparse.ArgumentParser()

    help_text = 'This tool returns information from AWS S3 buckets in a Amazon account'
    parser = argparse.ArgumentParser(description = help_text)

    # add long and short argument
    # parser.add_argument("--filter", "-f", help="Filter results by a given string")
    
    # add long and short argument
    parser.add_argument("--verbose", "-v", help="Verbose mode", action="store_true")
    
    # add long and short argument
    parser.add_argument("--bucket", "-b", help="Bucket name" )

    parser.add_argument("--unit", "-u", help="unit")
    
    # read arguments from the command line
    return parser.parse_args()

    # check for --xxx

HIGH_BUCKETS_COUNT_WARNING = 100
args = parsearguments()
s3 = boto3.resource('s3')
client = boto3.client('s3')
main()

# ---------------------------------------------------------------------------
# other code bellow
print("\nThis is the end, my only friend, the end...")

# asyncio.run(main())

# async def get_files(bucket):
#     print('{} requested files\n'.format(bucket.name))

    # files = bucket.objects.all()
    # for file in files:
    #     print(file.key, '{:,.0f}'.format(file.size/float(1<<20))+' MB') #, file.last_modified)


# async def get_files(bucket):
#     async for i in bucket.objects.all():
#         print(i.key)

# async def get_buckets():
#     for i in s3.buckets.all():
#         yield i

# async def main():
#     buckets = s3.buckets.all()
    
#     async for bucket in get_buckets():
#         print(bucket.name)
#         # get_files(bucket)

#         files = bucket.objects.all()
#         for file in files:
#             print(file.key, '{:,.0f}'.format(file.size/float(1<<20))+' MB') #, file.last_modified)
#         # print()

# loop = asyncio.get_event_loop()
# try:
#     loop.run_until_complete(main())
# finally:
#     loop.close()

    # print("Listing all buckets")

    # buckets = s3.buckets.all()


    # for bucket in buckets:
    #     files = bucket.objects.all()

    #     print('{} has {} files'.format(bucket.name, len(list(files))))
        
    #     for obj in files:
    #         print(obj.key, '{:,.0f}'.format(obj.size/float(1<<20))+' MB', obj.last_modified)
    #         # print(dir(obj))
    #     print('')

# except botocore.exceptions.NoCredentialsError:
#     print("Sorry, I can't find your credentials.")
# except botocore.exceptions.ClientError:
#     print("Your credentials are invalid, please check your credentials file (~/.aws/credentials)")

# except Exception as e:
#     print("Unexpected error occoured, can't list buckets right now\n{}".format(e))

# client = boto3.client('ce')

# response = client.get_cost_and_usage(TimePeriod={
#         'Start': '2019-09-01 00:00:00',
#         'End': '2019-09-14 00:00:00'
#     }
# )

# print(response)

    # bucket1 = s3.Bucket('brunomarques2')
    # print(bucket1.name)


# async def get_files(bucket):
#     # asyncio.sleep(random.randint(0,5))
#     await asyncio.sleep(2)
#     print(bucket)
#     files = bucket.objects.all()

#     # for file in files:
#     #     print(file.key, '{:,.0f}'.format(file.size/float(1<<20))+' MB') #, file.last_modified)

# async def main():
#     tasks = []

#     for bucket in s3.buckets.all():
#         tasks.append(asyncio.create_task(get_files(bucket)))

#     # print(tasks)

#     for task in tasks:
#         await task

awsls.py
# Amazon SDK
import boto3

# Amazon SDK exceptions
from boto3.exceptions import botocore

# Arguments parsing utility
import argparse

# System library to exit to console
import sys

# Datetime for date manipulation
from datetime import datetime, timedelta

# Pandas for data filtering and visualization
import pandas as pd

class cBucket:
    """ Class that represents a bucket in AWS. Contains name, creation date, files count, total files size, most recent file date modification, files count per storage types, cost and region
    """
    total_size_list = []

    def __init__(self, name):
        self.name = name
        self.creation_date = datetime(1, 1, 1).replace(tzinfo=None)
        self.files_count = 0
        self.files_size = 0
        self._last_modified = datetime(1, 1, 1).replace(tzinfo=None)
        self.storage_types = {}
        self.cost = 0.
        self.region = ''

    @property
    def last_modified(self):
        if self._last_modified == datetime(1, 1, 1).replace(tzinfo=None):
            return "-"
        else:
            return self._last_modified
    
    @last_modified.setter
    def last_modified(self, last_modified):
        self._last_modified = last_modified

def format_size(value):
    """ Convert bytes to other units, such as KB, MB, GB
    """
    bit_shift = {"B": 0,
            "kb": 7,
            "KB": 10,
            "mb": 17,
            "MB": 20,
            "gb": 27,
            "GB": 30,
            "TB": 40,}
    if args.unit in bit_shift:
        return "{:,.0f} {}".format(value / float(1 << bit_shift[args.unit]),args.unit) 
    return "{} bytes".format(value)

def format_percent(part, whole):
    """ format percentage of value
    """
    proportion = get_proportion(part, whole)
    return "{:.2f}%".format(proportion * 100)

def get_proportion(part, whole):
    """ Get proportion between part and whole, if whole is zero, returns 0 to be used as 0%
    """
    if whole == 0:
        return 0
    return float(part)/float(whole)

def print_buckets(buckets):
    """ Print buckets in different ways
    """
    data = []
    names = []

    # check if buckets list is empty in the fastest way
    # "total_size_list" is a shared property from class cBucket. It can be calculated inside the loop for each bucket, but it's much faster this way
    if buckets:        
        total_size = sum(buckets[0].total_size_list)

    for bucket in buckets:
        if total_size == 0:
            bucket.cost = 0
        else:
            bucket.cost = get_proportion(bucket.files_size, total_size) * float(total_cost)

        # "d" is an array with all bucket data formated and ready to be displayed
        d = [
            bucket.region,
            bucket.creation_date,
            bucket.files_count,
            format_size(bucket.files_size),
            format_percent(bucket.files_size, total_size),
            bucket.last_modified,
            bucket.cost ]
        data.append(d)
        names.append(bucket.name)

    # "features" dictionary with key:values that are the collumns in final table. Keys are also used in --sort argument
    features = {'region':'Region', 'creation':'Creation Date', 'files':'Files', 'real_size':'Size', 'size':'% Size', 'modified':'Last modified' , 'cost':'Cost'}
    dataframe = pd.DataFrame(data=data,columns=features.values(),index=names)

    if args.group:
        dataframe = dataframe.groupby('Region').sum()        
    elif args.regionfilter:
        dataframe = dataframe[dataframe.Region == args.regionfilter]
    
    if args.sort:
        if not args.group or (args.sort != 'creation' and args.sort != 'modified'):
            dataframe = dataframe.sort_values(features[args.sort], ascending=False)

    if dataframe.empty:
        print("No buckets match your criteria")
    else:
        print(dataframe)

def is_valid_file(f):
    """ Return a boolean indicating if file f is valid
    """
    result = False
    if f.storage_class == args.type or not(args.type):
        if str(args.filefilter) in str(f.key) or not(args.filefilter):
            result = True
    return result
    
def get_bucket_details(bucket):
    """ Get bucket details from amazon, create an populate a cBucket object and return it at the end
    """ 
    if args.verbose:        
            print('\tRetreving files in bucket {}'.format(bucket.name), end='...', flush=True)
    files = bucket.objects.all()
    if args.verbose:        
            print('done')
    
    last_modified_file = datetime(1, 1, 1).replace(tzinfo=None)
    total_files_size = 0
    valid_files_count = 0
    storage_types = {}

    for f in files:
        if is_valid_file(f):
            total_files_size+=f.size
            valid_files_count+=1

            last_change = f.last_modified.replace(tzinfo=None)
            if last_modified_file < last_change or last_modified_file=='':
                last_modified_file = last_change
            
            if (f.storage_class in storage_types): 
                storage_types[f.storage_class] += 1
            else: 
                storage_types[f.storage_class] = 1

    location_response = client.get_bucket_location( Bucket=bucket.name )
    location = location_response['LocationConstraint']

    coveo_bucket = cBucket(bucket.name)
    coveo_bucket.creation_date = bucket.creation_date.replace(tzinfo=None)
    coveo_bucket.files_count = valid_files_count
    coveo_bucket.files_size = total_files_size
    coveo_bucket.last_modified = last_modified_file.replace(tzinfo=None)
    coveo_bucket.storage_types = storage_types   
    coveo_bucket.region = location_response['LocationConstraint']
    coveo_bucket.total_size_list.append(total_files_size)

    return coveo_bucket

def get_y_n(text):
    """ Get Y or N from user with a friendly message. No input returns False (N)
    """
    valid_answers = [ 'y','n','Y','N', '' ]
    while True:
        data = input(text)
        if data not in valid_answers:
            continue
        elif data=='y' or data=='Y':
            return True
        else:
            return False

def get_buckets():
    """ Get all buckets from Amazon and return a filtered subset of them using command-line filters
    """
    global s3
    buckets = s3.buckets.all()
    filtered_buckets = []
    
    for bucket in buckets:
        if str(args.bucketfilter) in str(bucket.name) or not(args.bucketfilter):
            location = ''
            if args.regionfilter:
                location_response = client.get_bucket_location( Bucket=bucket.name )
                location = location_response['LocationConstraint']
            if location == args.regionfilter or not(args.regionfilter):
                filtered_buckets.append(bucket)
    
    return filtered_buckets

def parse_arguments():
    """ Parse arguments from command-line
    """
    parser = argparse.ArgumentParser()
    help_text = 'This tool returns useful information from AWS S3 bucket'
    parser = argparse.ArgumentParser(description = help_text)
    
    aws_regions = [ 'us-east-2',
        'us-east-1',
        'us-west-1',
        'us-west-2',
        'ap-east-1',
        'ap-south-1',
        'ap-northeast-3',
        'ap-northeast-2',
        'ap-southeast-1',
        'ap-southeast-2',
        'ap-northeast-1',
        'ca-central-1',
        'cn-north-1',
        'cn-northwest-1',
        'eu-central-1',
        'eu-west-1',
        'eu-west-2',
        'eu-west-3',
        'eu-north-1',
        'me-south-1',
        'sa-east-1',
        'us-gov-east-1',
        'us-gov-west-1']

    parser.add_argument("--verbose", "-v", help="Verbose mode", action="store_true")
    parser.add_argument("--group", "-g", help="Group by regions", action="store_true")    
    # parser.add_argument("--cost", "-c", help="Try to get the cost", action="store_true")    
    parser.add_argument("--bucketfilter", "-b", help="Bucket name filter" )
    parser.add_argument("--regionfilter", "-r", help="Region filter", choices=aws_regions)
    parser.add_argument("--filefilter", "-f", help="File filter")
    parser.add_argument("--type", "-t", help="Storage type", choices=[ 'STANDARD', 'REDUCED_REDUNDANCY', 'STANDARD_IA', 'ONEZONE_IA', 'INTELLIGENT_TIE' ] )
    parser.add_argument("--unit", "-u", help="Size unit", choices=[ 'kb', 'KB', 'mb', 'MB', 'gb', 'GB', 'TB'])
    parser.add_argument("--sort", "-s", help="Sort by feature (descending)", choices=[ 'region', 'creation', 'files', 'modified', 'size', 'cost' ])

    return parser.parse_args()

def get_total_cost():
    """ Get total storage cost from last day from Amazon. Can't be filtered
    """
    if args.bucketfilter or args.filefilter or args.regionfilter or args.type:
        print("WARNING: cost estimation isn't compatible with filtering. Cost results may not be reliabe.")

    today = datetime.today().date()
    yesterday = today-timedelta(days=2)
    try:
        ce = boto3.client('ce')
        cost_response = ce.get_cost_and_usage(TimePeriod={
                                                'Start': str(yesterday),
                                                'End': str(today)
                                            },
                                            Granularity='MONTHLY',
                                            Metrics=[ 'AmortizedCost']
                                        )
        c = cost_response['ResultsByTime'].pop()['Total']['AmortizedCost']['Amount']
    except Exception:
        print("WARNING: Can't get cost estimation from Amazon, cost will be zero")
        c = 0
    return c

def main():
    """ Main method that control global flow
    """
    try:
        buckets = get_buckets()
        buckets_count = len(buckets)
        
        global total_cost
        total_cost = get_total_cost()

        if args.verbose:        
            print('{} buckets found'.format(buckets_count))
            
        if buckets_count > HIGH_BUCKETS_COUNT_WARNING:
            allowed = get_y_n("WARNING: high number of buckets found ({}), it may take a while to get all data. Continue (y/N)?".format(buckets_count))
            if not(allowed):
                sys.exit()
    
        target_buckets = []
        for bucket in buckets:  
            coveo_bucket = get_bucket_details(bucket)
            target_buckets.append(coveo_bucket)
        
        print_buckets(target_buckets)

    except botocore.exceptions.NoCredentialsError:
        print("Sorry, I can't find your credentials file. Please double check the ~/.aws/credentials configuration file")
    except botocore.exceptions.ClientError as e:
        print("Client error, see details below:\n {}".format(e))
    except Exception as e:
        print("Unexpected error occoured, more details bellow\n{}".format(e))

# Set constants
HIGH_BUCKETS_COUNT_WARNING = 3

# Initialize global variables
total_cost = 0

# Parse command-line arguments
args = parse_arguments()

# Instantiates Amazon SDK resources
s3 = boto3.resource('s3')
client = boto3.client('s3')

# Main function
main()